flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/BufferedUpsertWriter.java

```diff
@@ -0,0 +1,209 @@
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import static org.apache.flink.streaming.connectors.kafka.table.DynamicKafkaRecordSerializationSchema.createProjectedRow;
+import static org.apache.flink.types.RowKind.DELETE;
+import static org.apache.flink.types.RowKind.UPDATE_AFTER;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+class BufferedUpsertWriter<WriterState> implements SinkWriter<RowData, Void, WriterState> {
+
+    private final SinkWriter<RowData, ?, WriterState> wrappedWriter;
+    private final WrappedContext wrappedContext = new WrappedContext();
+    private final ScheduledExecutorService scheduler =
+            Executors.newScheduledThreadPool(
+                    1, new ExecutorThreadFactory("upsert-kafka-sink-function"));
+    private final ScheduledFuture<?> scheduledFuture;
+    private final int batchMaxRowNums;
+    private final Function<RowData, RowData> valueCopier;
+    private final Map<RowData, Tuple2<RowData, Long>> reduceBuffer = new HashMap<>();
+    private final Function<RowData, RowData> keyExtractor;
+
+    private boolean closed = false;
+    private int batchCount = 0;
+    private volatile Exception flushException;
+
+    BufferedUpsertWriter(
+            SinkWriter<RowData, ?, WriterState> wrappedWriter,
+            DataType physicalDataType,
+            int[] keyProjection,
+            SinkBufferFlushMode bufferFlushMode,
+            @Nullable TypeSerializer<RowData> typeSerializer) {
+        checkArgument(bufferFlushMode != null && bufferFlushMode.isEnabled());
+        this.wrappedWriter = checkNotNull(wrappedWriter);
+        this.batchMaxRowNums = bufferFlushMode.getBatchSize();
+        long batchIntervalMs = bufferFlushMode.getBatchIntervalMs();
+
+        List<LogicalType> fields = physicalDataType.getLogicalType().getChildren();
+        final RowData.FieldGetter[] keyFieldGetters =
+                Arrays.stream(keyProjection)
+                        .mapToObj(
+                                targetField ->
+                                        RowData.createFieldGetter(
+                                                fields.get(targetField), targetField))
+                        .toArray(RowData.FieldGetter[]::new);
+        this.keyExtractor = rowData -> createProjectedRow(rowData, RowKind.INSERT, keyFieldGetters);
+        this.valueCopier = typeSerializer == null ? Function.identity() : typeSerializer::copy;
+
+        this.scheduledFuture =
+                this.scheduler.scheduleWithFixedDelay(
+                        () -> {
+                            synchronized (BufferedUpsertWriter.this) {
+                                if (!closed) {
+                                    try {
+                                        flush();
+                                    } catch (Exception e) {
```

Comment:
You probably need to add InterruptionException to ProcessingTimeCallback#onProcessingTime, which I forgot to implement ;)